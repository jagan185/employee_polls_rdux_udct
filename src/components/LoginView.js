import React, { useState } from "react";
import { connect } from "react-redux";
import { useLocation, useNavigate } from "react-router-dom";
import { login } from "../actions/authorizeUser.js";

const LoginView = ({ userIds, users, dispatch }) => {
  const [optionSelected, setOptionSelected] = useState("");
  const [success, setSuccess] = useState(false);
  const [error, setError] = useState(false);

  const navigate = useNavigate();
  const location = useLocation();

  const handleChange = (event) => {
    event.preventDefault();
    setOptionSelected(event.target.value);

    if (event.target.value !== "" && event.target.value !== "none") {
      dispatch(login(event.target.value));
      navigate(location?.state?.location);
    } else if (event.target.value === "none") {
      dispatch(login(""));
      navigate("/login");
    }
  };

  /**
   * @description mostly used to test the login functionality, onchange actullay logs in and navigates too
   */
  const handleSubmit = (event) => {
    event.preventDefault();
    console.log("submitedd:::");
    setSuccess(true);
    setError(false);
    //navigate("/");
  };

  return (
    <section className="h-full gradient-form bg-gray-200 md:h-screen">
      <div className="container py-12 px-6 h-full">
        {success && <h1 data-testid="successMsgEle">Logged In!!</h1>}
        {error && (
          <h1 data-testid="errorMsgEle">
            Couldn't log in, select a user.
          </h1>
        )}
        <div className="flex justify-center items-center flex-wrap g-6 text-gray-800" key="flexDiv234r">

          <div className="lg:flex lg:flex-wrap g-0" key="formDiv23423">

            <form onSubmit={handleSubmit} key="formItem23423">
              <p className="mb-4">Please login to your account</p>
              <div className="mb-4" key="selectDiv32423">
                <select
                  name="users"
                  onChange={handleChange}
                  data-testid="userDropdownEle"
                  className="form-select mt-3"
                  defaultValue={optionSelected}
                  aria-label="Default select example"
                  key="selectItem2343"
                >
                  <option value="none" key="noneOption324333"></option>
                  {userIds.map((id) => {
                    return (
                        <option value={id} key={id}>
                          {users[id].name}
                        </option>
                    );
                  })}
                  ;
                </select>
              </div>
              <div className="text-center pt-1 mb-12 pb-1">
                <button
                  className="inline-block px-6 py-2.5 hover:text-white font-medium text-xs leading-tight uppercase rounded shadow-md hover:bg-blue-700 hover:shadow-lg focus:shadow-lg focus:outline-none focus:ring-0 active:shadow-lg transition duration-150 ease-in-out w-full mb-3"
                  type="button"
                  data-mdb-ripple="true"
                  data-mdb-ripple-color="light"
                  data-testid="submitBtnEle"
                >
                  Log in
                </button>
              </div>
            </form>
          </div>
        </div>
      </div>
    </section>
  );
};
const mapStateToProps = (state) => {
  const users = state.users;

  return {
    userIds: Object.keys(users),
    users,
  };
};

export default connect(mapStateToProps)(LoginView);
