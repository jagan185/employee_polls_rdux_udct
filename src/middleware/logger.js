/**
 * logger is a function that takes store as input and returns a function that takes next as input and returns a function that takes action as input logs action and calls next function passing in action, which returns result, this result is returned by logger
 * @param {*} store 
 * @returns 
 */
const logger = (store) => (next) => (action) => {
    console.group(action.type);
    console.log("action :: ", action);
    const result = next(action);
    console.log("state from store :: ", store.getState());
    console.groupEnd();
    return result;
  };

  export default logger;