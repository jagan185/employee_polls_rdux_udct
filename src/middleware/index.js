import thunk from "redux-thunk";
import { applyMiddleware } from "redux";
import logger from "./logger.js";
/**
 * combine middlewares here,
 * middlewares are applied in the order they are passed to applyMiddleware
 */
export default applyMiddleware(thunk, logger);