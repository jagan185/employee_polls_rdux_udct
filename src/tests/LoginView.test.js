/**
 * @description unit tests for LoginView.js
 */
import { Provider } from "react-redux";
import { legacy_createStore } from "redux";
import { MemoryRouter } from "react-router-dom";
import { render, screen, fireEvent } from "@testing-library/react";
import '@testing-library/jest-dom';

import reducer from "../reducers";
import middleware from "../middleware";
import LoginView from "../components/LoginView.js";



const store = legacy_createStore(reducer, middleware);

describe("LoginView", () => {
  it("will log in if a user is selected.", () => {
    render(
      <MemoryRouter>
        <Provider store={store}>
          <LoginView />
        </Provider>
      </MemoryRouter>
    );
    
    let username = screen.getByTestId("userDropdownEle");
    fireEvent.change(username, { target: { value: "sarahedo" } });
    console.log('tst usrename : '+username);
    let submitButton = screen.getByTestId("submitBtnEle");
    fireEvent.click(submitButton); //this won't work since the onSubmit handler is on form and not the submit
    //submit the form with the selected username
    fireEvent.submit(username);
    console.log('::: elee :'+screen.getByTestId("successMsgEle"));
    expect(screen.getByTestId("successMsgEle")).toBeInTheDocument();
    expect(screen.queryByTestId("errorMsgEle")).not.toBeInTheDocument();
  });

  it("LoginView component rendered without changes should match snapshot", () => {
    let { component } = render(
      <MemoryRouter>
        <Provider store={store}>
          <LoginView />
        </Provider>
      </MemoryRouter>
    );

    expect(component).toMatchSnapshot();
  });
});