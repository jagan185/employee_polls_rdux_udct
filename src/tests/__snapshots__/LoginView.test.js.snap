// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginView LoginView component rendered without changes should match snapshot 1`] = `undefined`;
